File: da-hyun/week_1/Cpp/Baekjoon_1189.cpp
//
//  Baekjoon_1189.cpp
//  algorithmStudy
//
//  Created by TeamGonom on 8/3/24.
//
#include <iostream>
#include <algorithm>
#include <cmath>

using namespace std;
int dx[4] = {-1, 0, 1, 0};
int dy[4] = {0, 1, 0, -1};
int recursive(int x, int y, int depth, bool** board, bool** visited, int N, int M, int K){
    if((x == 0 && y == M-1) || depth == K){
        if((x == 0 && y == M-1) && depth == K){
            return 1;
        }
        return 0;
    }
    visited[x][y] = true;
    int tmpAnswer = 0;
    for(int i=0;i<4;i++){
        int nx = x + dx[i], ny = y + dy[i];
        if(nx < 0 || ny < 0 || nx>=N || ny >= M || visited[nx][ny] || !board[nx][ny]) continue;
        tmpAnswer += recursive(nx, ny, depth+1, board, visited, N, M, K);
    }
    visited[x][y] = false;
    return tmpAnswer;
}

int main(){
    int N, M, K;
    cin >> N >> M >> K;
    bool ** visited, ** map;
    map = new bool*[N];
    visited = new bool*[N];
    for(int i=0;i<N;i++){
        map[i] = new bool[N];
        visited[i] = new bool[N];
        for(int j=0;j<M;j++){
            char A;
            cin >> A;
            map[i][j] = (A == '.');
            visited[i][j] = false;
        }
    }
    cout << recursive(N-1, 0, 1, map, visited, N, M, K);
    return 0;
}

File: da-hyun/week_1/Python/Baekjoon_1189.py
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

def recursive(x, y, depth, board, visited, R, C, K):
    # print((x, y))
    visited[x][y] = True
    if(x, y) == (0, C-1) or depth == K:
        visited[x][y] = False
        if((x,y) == (0,C-1) and depth == K): return 1
        return 0
    tmpAnswer = 0
    for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        if(0 <= nx < R and 0 <= ny < C and not visited[nx][ny] and board[nx][ny]):
            tmpAnswer += recursive(nx, ny, depth + 1, board, visited, R, C, K)
    visited[x][y] = False
    return tmpAnswer

def main():
    R, C, K = map(int, input().split())
    board = [list(map(lambda x: x=='.', input())) for _ in range(R)]
    visited = [[False]*C for _ in range(R)]
    print(recursive(R-1, 0, 1, board, visited, R, C, K))

if __name__ == "__main__":
    main()